### PR TITLE
Drop gtk and use gdk-pixbuf

### PR DESCRIPTION
**Pull request by bebehei**
Removes the neccessity of gdk3.

From #334:

> As it is right now it runs in ~20-30 microseconds and according to my tests when converting it to a png and back goes up to ~200-250 microseconds.
> 
> While that's obviously going to vary from CPU to CPU but a 10x overhead is a significant slowdown and I further discovered that this function gets called every time there's a focus change, that's a lot of needless CPU time.

That's correct. It's unacceptable to have a 10x slowdown. Additionally I think also it is bad spending on every focus change cpu power for the exact same calculation, when it's ok to save it in memory. So why should we spend even 20-30 usec on it?

With this PR, dunst generates the `cairo_surface_t` for the icon now in `notification_init` and saves it as a part of the notification.

> I pushed a proof of concept to a new branch but I really don't like how this is implemented (especially having a separate struct just to pass the buffer pointer and size)

Using `GByteArray` from glib fixes this quite nicely.

---

The code is currently not the best. There are still a few TODO marks left in the commits, but I want to get general feedback before fine-tuning this PR.

Not using `gdk_cairo_set_source_pixbuf` because of the dependency to gtk3, might be a bit counterintuitive in the code, but I think having just because of this function call >160MB dependencies, while dunst is 160KB big it's ok to use a workaround here. KIM, why dunst is mainly used: 

> Dunst is a _lightweight_ repl....

And I think it's ridiculous to pretend it's lightweight, while the deps are factor 1000 bigger.

### CODE REVIEW

**Comment by tsipinakis** [comment]
Wow, I just came back and was about to start looking into ways of refactoring that part to avoid multiple calls to that function but looks like you beat me to it, thanks for yet another awesome contribution!

This also makes merging the `refactor-drawing` branch a pain to the point where I'll probably start over but I dug myself into this mess for delaying working on it for so long.

Review coming up shortly.

**Comment by bebehei** [comment]
> Review coming up shortly.

Wait with a review. I wanted have a green light and I got that now, thanks. I'll ask you again later for a review.

**Comment by bebehei** [comment]
@tsipinakis I've overhauled all undecided bits of the code. If you'd like to review it now, that'd be great!

File: src/x11/x.c [hunk 1]
```diff
@@ -298,47 +298,9 @@ static colored_layout *r_init_shared(cairo_t *c, notification *n)
...
-                cl->icon = NULL;
-        }
-
-        if (cl->icon && cairo_surface_status(cl->icon) != CAIRO_STATUS_SUCCESS) {
-                cairo_surface_destroy(cl->icon);
-                cl->icon = NULL;
+        if (settings.icon_position != icons_off) {
+                cl->icon = n->icon_surface;
+                n->icon_surface = cairo_surface_reference(n->icon_surface);
         }
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/x11/x.c]: I don't understand your objection. If cl->icon is uninitialized, nobody cares about this and we overwrite the pointer with n->icon_surface.

[the remark this answers is not in the document]
Comment (tsipinakis) [reply src/x11/x.c]: > If cl->icon is uninitialized, nobody cares about this

`calculate_dimentions` in x.c does, the only check for icons is if `cl->icon` is `NULL` or not, which causes a segfault if it's uninitialized.

[the remark this answers is not in the document]
Comment (bebehei) [reply src/x11/x.c]: Thanks, I get it.

**Comment by tsipinakis** [comment]
As per my earlier review I've moved most pango/cairo function into `draw.c` in the `refactor-drawing` branch so I'm changing the target branch to that, it's still not in a merge-able form but it's close enough to unblock this.

I'd prefer if all graphical related calls/functions stay as much as possible within the relevant files (`draw.c`,`x.c`, the new `icon.c` file etc).

**Comment by bebehei** [comment]
So, I came back to this PR now, as the others for 1.3 are mostly solved.

> I don't like is coupling the dbus/notification interface with the drawing one, I believe it's best to keep things separate. For that reason we have a separate struct in x.c for storing objects related to rendering the notification: `colored_layout`.

I took the time now to understand, what I've done. Man, I've done shit in this PR. I really haven't understood our x11 lump at that point.

But I still don't understand one point: I can't find a line, where these colored_layouts are saved.

> Looking at the drawing procedure this issue runs far deeper than the icons, x_win_draw which generates the layout for each notification gets called very frequently. It's one of the issues I tried to address in drawing-refactor.

So is it right, that these layouts aren't saved, generated freshly with every `x_win_draw` call?

If so, my PR does performance much better ;-) but the code quality in this PR still sucks.

---

So:

We should actually have crosslinks (not just single from cl to n) between `notification` and `colored_layout` struct. 

Somewhere in `x_win_draw` or `r_create_layouts`, we should check if `n->colored_layout == NULL` and create a new layout, which we save in there and the next time we have it already.

`notification_free` also calls `colored_layout_free`  on its own (newly introduced) `colored_layout` field.

**Comment by bebehei** [comment]
Ok, I've overhauled most bits. `notification` links now to `colored_layout` and does not link to cairo specifics. On the one hand side, that's ok. But now they're cross referencing themselfes, which requires special prototyping.

On the speed issue: I can see a clear decrease of execution time of `x_win_draw`. When waking up dunst because of some X11 event, it takes a fraction of time (~2000 vs ~750 clocks).

**Comment by bebehei** [comment]
I changed the branch back to `master`, as the last commit from `refactor-drawing` is nearly unportable to current master.

I also have taken advance and pushed the `colored_layout`,specifics to `draw.h`.

**Comment by bebehei** [comment]
@tsipinakis to bring this further: I actually would drop the current last two commits (freeing colored layout upon history push, linking notificaiton and colored layout).

 I don't know how to do these better.

You're absolutely right with your speed concern. I also don't think that it's good to buy dropping gtk3 with additional CPU usage. But I think its temporarily ok.

I'd propose to drop solely GTK3 requirement here, do not any optimisations, release dunst 1.3 regardless if it's actually using more CPU time or not. And after the release of 1.3 then refactor and optimise the whole drawing stuff for dunst 1.4 in calmness.

I think it's the only solution to get this situation handled and not deadlock it for another two months. I've got the feeling we're stepping on each other's feet. You want to start refactor drawing, but require this thing to get merged, while I want to merge this, but require some refactored stuff.

**Comment by tsipinakis** [comment]
> I'd propose to drop solely GTK3 requirement here, do not any optimisations, release dunst 1.3 regardless if it's actually using more CPU time or not. And after the release of 1.3 then refactor and optimise the whole drawing stuff for dunst 1.4 in calmness.

I agree about merging this without the optimizations but not about releasing 1.3, we're not in a hurry to release and in my opinion we should wait until we at least make it so that it's not reimporting the image every time the focus changes, it's a bit too much. 

> I think it's the only solution to get this situation handled and not deadlock it for another two months. I've got the feeling we're stepping on each other's feet. You want to start refactor drawing, but require this thing to get merged, while I want to merge this, but require some refactored stuff.

Agreed, in any other situation I'd say "lets merge it in a separate branch and work there" but the drawing refactor is a huge chunk of work that will conflict with anything and I'd take some increased CPU usage and maybe instability due to the refactor in master over resolving endless conflicts anyday.

**Comment by bebehei** [comment]
> I agree about merging this without the optimizations but not about releasing 1.3, we're not in a hurry to release and in my opinion we should wait until [...]

Well, we're not in a hurry. But the milestone for 1.3 is cleared after this (and #428). I don't like to wait for artificial work. Dangling unreleased commits on master is no one's favor.

>  [...] we at least make it so that it's not reimporting the image every time the focus changes, it's a bit too much.

I've benchmarked it again. [But I look at `x_win_draw` in general and not specifically the icon conversion.](https://github.com/bebehei/dunst/commits/drop-gtk-dbg). Those numbers are IMHO more meaningful.

On my system it reports ~2000 microseconds execution time for `x_win_draw` on master branch and ~3000 microseconds execution time on `drop-gtk`. I agree it's 50% growth. But it's a millisecond difference per focus change.... How often do you change your focus while a notification is displayed?

Keep in mind: We're talking here about a change, which lets dunst decrease its dependency footprint by factor 1000 while increasing cputime by factor 1.5.

Based on those numbers, I don't see why we should delay releasing it.

File: src/x11/x.c [hunk 2]
```diff
@@ -908,9 +910,17 @@ gboolean x_mainloop_fd_dispatch(GSource *source, GSourceFunc callback, gpointer
...
-                case PropertyNotify:
                         wake_up();
                         break;
+                case PropertyNotify:
+                        scr = get_active_screen();
+                        /* Ignore PropertyNotify, when we're still on the
+                         * same screen. PropertyNotify is only neccessary
+                         * to detect a focus change to another screen
+                         */
+                        if(scr->scr != xctx.cur_screen)
```
Comment (tsipinakis) [inline src/x11/x.c]: You can limit when this runs even more by checking that `settings.f_mode != FOLLOW_NONE`

**Comment by bebehei** [comment]
> Edit: I just read the IRC memo why you rebased the entire PR, that's weird.

Well, I also did a force push, because GitHub dismissed your review when I pushed these additional commits.

> I'd implement it myself now but collaboration inside PRs

moving git commits isn't a problem a problem for me. It'll just become a problem, if you go on and make a commit by ~600lines like on the drawing refactoring. There it's not possible to compare the diffs and see, that the diff is actually correct. So either work on somewhere else and tell me, you have additional work. Or push it directly on my branch. I always "Allow edits from maintainers" in my PRs. Only if you do this, you should make a comment and tell me about this in the PR. I'll likely dismiss those "XY has pushed x commits to PR".

File: src/x11/x.c (hunk 2 above)
Comment (bebehei) [reply src/x11/x.c]: Handled now.

**Comment by bebehei** [comment]
> it dawned on me: Most icons are in svg format and cairo has a native cairo_svg_surface_create funciton. It would be much better performance wise if we use that for all svg images and fallback to gdk_pixbuf -> png -> cairo for everything else.

Full ACK. But the [docs to cairo_svg_surface_create](https://www.cairographics.org/manual/cairo-SVG-Surfaces.html#cairo-svg-surface-create) read this:

    Creates a SVG surface of the specified size in points to be written to filename.

This looks quite bogus to me. I have consulted #cairo IRC channel.

While writing, benzea already answered me:

```
<bebehei> Hi guys. Could somebody please illuminate me? There is cairo_svg_surface_create. I've got the expectation, that this function creates a cairo_surface_t by reading an svg file. But the docs read completely different. "Creates a SVG surface [...] to be written to filename." Could it be, that the docs are wrong?
<bebehei> https://www.cairographics.org/manual/cairo-SVG-Surfaces.html#cairo-svg-surface-create
<benzea> bebehei: cairo can only render to SVG not load SVGs
<benzea> bebehei: you can use librsvg if you need to load/render SVGs
<bebehei> benzea: Thank you. Just to understand this fully. How would I use cairo_svg_surface_create to export a surface to an SVG image? At this point the signature does not yet make sense to me.
<benzea> bebehei: if you create an SVG surface using cairo_svg_surface_create you can draw to it normally after creating a cairo context for it
<bebehei> benzea: And when drawing onto that surface, automatically everything also gets written into this file, right?
<benzea> yup, you should get a proper SVG file after you finish drawing to it
<bebehei> benzea, thank you very much.
<benzea> I guess you'll need to destroy the surface before for the file to be written completely
```

So, sadly, no. We can't easily import an SVG file with cairo.

**Comment by tsipinakis** [comment]
> So, sadly, no. We can't easily import an SVG file with cairo.

That's unfortunate, the wording of the docs had me confused there but I guess this will have to do.

**Comment by bebehei** [comment]
Thanks @tsipinakis for the final Approval.

**Pull request merged**